base/compiler/ssair/passes.jl

```diff
@@ -781,6 +783,7 @@ function sroa_pass!(ir::IRCode)
                 continue
             end
             if !isempty(new_preserves)
+                compact[idx] = nothing
                 compact[idx] = form_new_preserves(stmt, preserved, new_preserves)
```

Comment:
Since `compact[idx] === stmt`, we first need to first set `compact[idx]` to `nothing`, because in-place modification can lead to miscounts by `setindex`.